File: backend/evtx/permissions.go
// Package evtx owns EVTX/Windows event-log ingestion for rohy: privilege
// checks, streaming file parsing, live event-log access, and normalization into
// the persistence layer's Event model. This file holds the platform-agnostic
// permission logic; platform-specific detection lives in permissions_windows.go
// and permissions_other.go.
package evtx

import (
	"fmt"
	"strings"

	"rohy/backend/consts"
)

// PermissionStatus reports the current process privilege state. It is JSON-tagged
// because it is surfaced to the frontend permissions store via the API layer.
type PermissionStatus struct {
	Platform      string `json:"platform"`
	Elevated      bool   `json:"elevated"`
	Administrator bool   `json:"administrator"`
}

// AccessDecision is the result of evaluating whether a set of requested channels
// can be read under a given permission status. When Needed is true the caller
// should surface Message via the Material snackbar and fall back to file ingest.
type AccessDecision struct {
	Needed          bool     `json:"needed"`
	BlockedChannels []string `json:"blocked_channels"`
	Message         string   `json:"message"`
}

// RequiresElevation reports whether reading the named channel needs administrator
// rights. Matching is case-insensitive.
func RequiresElevation(channel string) bool {
	for _, c := range consts.ElevatedChannels {
		if strings.EqualFold(c, channel) {
			return true
		}
	}
	return false
}

// EvaluateAccess determines whether any requested channel is blocked under status
// and, if so, builds the warning message. Channels that do not require elevation,
// or all channels when the process is elevated, produce a zero-value decision
// (Needed == false).
func EvaluateAccess(channels []string, status PermissionStatus) AccessDecision {
	var blocked []string
	for _, ch := range channels {
		if RequiresElevation(ch) && !status.Elevated {
			blocked = append(blocked, ch)
		}
	}
	if len(blocked) == 0 {
		return AccessDecision{}
	}
	return AccessDecision{
		Needed:          true,
		BlockedChannels: blocked,
		Message:         fmt.Sprintf(consts.MsgElevationRequired, strings.Join(blocked, ", ")),
	}
}

File: frontend/src/stores/permissions.js
// Permissions store (P5.2, closes P1's frontend items). Holds the latest privilege
// snapshot and the access decision for the protected channels, refreshed on demand.
// A blocked decision is surfaced through the snackbar.
import { writable } from 'svelte/store';
import * as api from '../lib/api/index.js';
import { PROTECTED_CHANNELS } from '../lib/consts/index.js';
import { snackbar } from './snackbar.js';

function create() {
  const { subscribe, set, update } = writable({
    checked: false,
    status: { platform: '', elevated: false, administrator: false },
    decision: { needed: false, blocked_channels: [], message: '' },
    error: null,
  });

  /** Refresh privilege state + protected-channel access. Warns via snackbar if blocked. */
  async function refresh() {
    try {
      const status = await api.checkPermissions();
      const decision = await api.evaluateAccess(PROTECTED_CHANNELS);
      set({ checked: true, status, decision, error: null });
      if (decision && decision.needed && decision.message) {
        snackbar.warn(decision.message);
      }
      return { status, decision };
    } catch (err) {
      update((s) => ({ ...s, checked: true, error: String(err && err.message ? err.message : err) }));
      return null;
    }
  }

  return { subscribe, refresh };
}

export const permissions = create();
